### PR TITLE
Feat/segmentation fec

### LINKED ISSUE #4330: Enhance reliability in message segmentation
# Problem

Our current segmentation mechanism divides messages into multiple segments and dispatches these as distinct Waku messages to clients. A critical limitation of this approach is the lack of a recovery mechanism for scenarios where one or more segments are lost during transmission.

It is essential to improve the reliability of our message segmentation technique, with an emphasis on conserving network bandwidth. Towards achieving this, I propose a two-tiered solution:
1) Integration of Parity Data: Implementing an erasure coding strategy (see: https://en.wikipedia.org/wiki/Erasure_code). By adding parity data to the transmitted message, we can enable the reconstruction of the original data even when certain segments are lost. This method leverages redundancy to enhance fault tolerance without excessively increasing bandwidth usage.

2) Client-Driven Segment Redelivery Requests:  This approach enables clients to identify and request the re-sending of any segments they have missed. Considering that typically only a few segments fail to arrive, this method should not significantly impact bandwidth usage.

### PR DESCRIPTION
**Pull request by osmaczko**
Introduce forward error correction at the segmentation layer. The change is fully compatible:
- Old clients will be able to read messages from new clients.
- New clients will be able to read messages from old clients.

Closes: https://github.com/status-im/status-go/issues/4330

### CODE REVIEW

File: protocol/common/message_segmentation.go [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+		}
+
+		copy.Payload = chunkPayload
+		copy.PowTarget = calculatePoW(chunkPayload)
+		return copy, nil
+	}
+
+	entireMessageHash := crypto.Keccak256(newMessage.Payload)
+	payloadSize := len(newMessage.Payload)
+	segmentsCount := int(math.Ceil(float64(payloadSize) / float64(maxSegmentSize)))
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: Maybe add `maxSegmentSize > 0` check in the beginning not to accidentally divide by zero here? 🤔

File: protocol/common/message_segmentation.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+		copy := &types.NewMessage{}
+		err := copier.Copy(copy, newMessage)
+		if err != nil {
+			return nil, err
+		}
+
+		copy.Payload = chunkPayload
+		copy.PowTarget = calculatePoW(chunkPayload)
+		return copy, nil
+	}
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: Does this func need to be defined here? Maybe keep it outside?

File: protocol/common/message_segmentation.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+	payloads := make([][]byte, segmentsCount+paritySegmentsCount)
+	segmentMessages := make([]*types.NewMessage, segmentsCount)
+
+	for start, index := 0, 0; start < payloadSize; start += maxSegmentSize {
+		end := start + maxSegmentSize
+		if end > payloadSize {
+			end = payloadSize
+		}
+
+		chunk := newMessage.Payload[start:end]
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: `chunk` is actually a `segment`, right? 
Using 2 terms is a bit confusing

Comment (igor-sirotin) [reply protocol/common/message_segmentation.go]: And I guess `payloads` would become `segments` then. Or `chunks` worst case 😄

File: protocol/common/message_segmentation.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+			return nil, err
+		}
+
+		payloads[index] = chunk
+		segmentMessages[index] = segmentMessage
+		index++
+	}
+
+	if paritySegmentsCount == 0 || segmentsCount+paritySegmentsCount > segmentsMaxCount {
+		return segmentMessages, nil
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: This check seems incorrect to me 🤔 

If we exceeded `segmentsMaxCount`, why do we just return `segmentMessagres`?
Or is this limit only about reed-solomon encoding?

File: protocol/common/message_segmentation.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+	message.TransportLayer.Payload = entirePayload.Bytes()
+
+	return nil
+}
+
+// SegmentationLayerV2 is capable of reconstructing the message from both complete and partial sets of data segments.
+// It has capability to perform forward error correction.
+func (s *MessageSender) handleSegmentationLayerV2(message *v1protocol.StatusMessage) error {
+	logger := s.logger.With(zap.String("site", "handleSegmentationLayerV2")).With(zap.String("hash", types.HexBytes(message.TransportLayer.Hash).String()))
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go:227]: There's a better way for this, we can use `Named`:
```suggestion
	logger := s.logger.Named("handleSegmentationLayerV2").With(zap.String("hash", types.HexBytes(message.TransportLayer.Hash).String()))
```

File: protocol/common/message_segmentation.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+		SegmentMessage: &protobuf.SegmentMessage{},
+	}
+	err := proto.Unmarshal(message.TransportLayer.Payload, segmentMessage.SegmentMessage)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal SegmentMessage")
+	}
+
+	logger.Debug("handling message segment", zap.String("EntireMessageHash", types.HexBytes(segmentMessage.EntireMessageHash).String()),
+		zap.Uint32("Index", segmentMessage.Index), zap.Uint32("SegmentsCount", segmentMessage.SegmentsCount),
+		zap.Uint32("ParitySegmentIndex", segmentMessage.ParitySegmentIndex), zap.Uint32("ParitySegmentsCount", segmentMessage.ParitySegmentsCount))
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: I'd much appreciate if each argument was on a separate line. It's much easier to read, and it's also easier to modify later (git diffs will be clearer).

But not a big deal, just some style 🧂 

```suggestion
	logger.Debug("handling message segment", 
	    zap.String("EntireMessageHash", types.HexBytes(segmentMessage.EntireMessageHash).String()),
		zap.Uint32("Index", segmentMessage.Index), 
		zap.Uint32("SegmentsCount", segmentMessage.SegmentsCount),
		zap.Uint32("ParitySegmentIndex", segmentMessage.ParitySegmentIndex),
		zap.Uint32("ParitySegmentsCount", segmentMessage.ParitySegmentsCount))
```

File: protocol/common/message_segmentation.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+	// First segment message must not be a parity message.
+	if firstSegmentMessage.IsParityMessage() || len(segments) != int(firstSegmentMessage.SegmentsCount) {
+		return ErrMessageSegmentsIncomplete
+	}
+
+	payloads := make([][]byte, firstSegmentMessage.SegmentsCount+lastSegmentMessage.ParitySegmentsCount)
+	payloadSize := len(firstSegmentMessage.Payload)
+
+	restoreUsingParityData := lastSegmentMessage.IsParityMessage()
+	if restoreUsingParityData {
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: Do you mind inverting this condition to make the shortest `if` branch go first?

File: protocol/common/message_segmentation.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+					// Ensure last segment is aligned to payload size, as it is required by reedsolomon.
+					payloads[segment.Index] = make([]byte, payloadSize)
+					copy(payloads[segment.Index], segment.Payload)
+					lastNotParitySegmentPayload = segment.Payload
+				} else {
+					payloads[segment.Index] = segment.Payload
+				}
+			} else {
+				payloads[firstSegmentMessage.SegmentsCount+segment.ParitySegmentIndex] = segment.Payload
+			}
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]:
```suggestion
			if segment.IsParityMessage() {
				payloads[firstSegmentMessage.SegmentsCount+segment.ParitySegmentIndex] = segment.Payload
				continue
			}
		    if segment.Index != firstSegmentMessage.SegmentsCount-1 {
			    payloads[segment.Index] = segment.Payload
			    continue
		    }
			// Ensure last segment is aligned to payload size, as it is required by reedsolomon.
			payloads[segment.Index] = make([]byte, payloadSize)
			copy(payloads[segment.Index], segment.Payload)
			lastNotParitySegmentPayload = segment.Payload
```

File: protocol/common/message_segmentation.go [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+	message.TransportLayer.Payload = entirePayload.Bytes()
+
+	return nil
+}
+
+func (s *MessageSender) CleanupSegments() error {
+	weekAgo := time.Now().AddDate(0, 0, -7).Unix()
+	monthAgo := time.Now().AddDate(0, -1, 0).Unix()
+
+	err := s.persistence.RemoveMessageSegmentsOlderThan(weekAgo)
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation.go]: Hmmm, but we could fetch the missing segments later from store nodes, couldn't we?

File: protocol/common/message_segmentation_test.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+		nil,
+		s.logger,
+		FeatureFlags{},
+	)
+	s.Require().NoError(err)
+
+	s.testPayload = make([]byte, 1000)
+	for i := 0; i < 1000; i++ {
+		s.testPayload[i] = byte(i)
+	}
```
Comment (igor-sirotin) [inline protocol/common/message_segmentation_test.go]: Maybe move this to `SetupSuite`?

File: protocol/common/message_sender_test.go [hunk 11]
```diff
@@ -361,3 +364,53 @@ func (s *MessageSenderSuite) TestHandleSegmentMessages() {
...
+	s.Require().True(ok)
+
+	s.Require().ElementsMatch(data[:10], dataCopy[:10])
+}
+
+func (s *MessageSenderSuite) TestDummy() {
+	for i := 0; i < 100; i++ {
+		fmt.Println("segments: ", i, ", parity: ", int(math.Floor(float64(i)/8)))
+	}
+}
```
Comment (igor-sirotin) [inline protocol/common/message_sender_test.go]: leftover?

**Review by igor-sirotin — APPROVED** [review]
Nice and so clean! Thank you ❤️

File: protocol/common/message_segmentation.go [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+	segmentsParityRate = 0.125
+	segmentsMaxCount   = 256
+)
+
+type SegmentMessage struct {
+	*protobuf.SegmentMessage
+}
+
+func (s *SegmentMessage) IsValid() bool {
+	return s.SegmentsCount >= 2 || s.ParitySegmentsCount > 0
```
Comment (jrainville) [inline protocol/common/message_segmentation.go:36]: Should the 2 be in a constant? Is this something that will be configurable or something?

File: protocol/common/message_segmentation.go (hunk 9 above)
Comment (jrainville) [reply protocol/common/message_segmentation.go]: Store nodes store up to 30 days, but we fetch only 7-9 days, so it works. But Maybe using a constant used in both would make more sense.

Also, does this clean up work on all segments or only the orphan ones? ie when a message is completed (we got all needed segments), are the segments kept 7 days or are they cleaned up automatically after?

edit: I just saw the `RemoveMessageSegmentsCompletedOlderThan(monthAgo)`. Shouldn't we clean up completed segments as soon as we complete why do we keep them?

**Review by jrainville — APPROVED** [review]
Good job, so small questions

File: protocol/common/message_segmentation.go (hunk 3 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go]: Yeah, a bit confusing. I improved the naming, hope to less confusing one.

File: protocol/common/message_segmentation.go (hunk 4 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go]: Renamed and added extra comment.

File: protocol/common/message_segmentation.go (hunk 5 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go:227]: It doesn't work with production logger we have configured. It only works with debug (test) loggers. I didn't dig to find the why.

File: protocol/common/message_segmentation.go (hunk 2 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go]: Extracted.

File: protocol/common/message_segmentation.go (hunk 1 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go]: Seems too defensive for me :P

File: protocol/common/message_segmentation.go (hunk 8 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go]: I prefer the way it is now, if you don't mind. Even though the indentation level is deeper, the logic reads better to me.

File: protocol/common/message_segmentation_test.go (hunk 10 above)
Comment (osmaczko) [reply protocol/common/message_segmentation_test.go]: Didn't know such API exists, thanks.

File: protocol/common/message_sender_test.go (hunk 11 above)
Comment (osmaczko) [reply protocol/common/message_sender_test.go]: Yeah, removed, thanks

File: protocol/common/message_segmentation.go (hunk 12 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go:36]: We don't create segments if the message fits within the limit; the segment count must be at least two. The condition will never change. Creating a constant for that may be overkill.

File: protocol/common/message_segmentation.go (hunk 9 above)
Comment (osmaczko) [reply protocol/common/message_segmentation.go]: > Hmmm, but we could fetch the missing segments later from store nodes, couldn't we?

That's right. I assumed that if one couldn't fetch the full message for a week, then the missing segment would never come. However, I overlooked the case when one goes on a two-week vacation and then fetches from store nodes. I changed both cleanups to a month interval.

> Also, does this clean up work on all segments or only the orphan ones? ie when a message is completed (we got all needed segments), are the segments kept 7 days or are they cleaned up automatically after?

They are cleaned up automatically after, see: `CompleteMessageSegments`.

> edit: I just saw the RemoveMessageSegmentsCompletedOlderThan(monthAgo). Shouldn't we clean up completed segments as soon as we complete why do we keep them?

The name is a bit unfortunate. It cleans up only the table with hashes, not the message segments themselves. There is a table named `message_segments_completed` that holds the hash of the completed message and the `timestamp` of when it was completed. This is to ignore segments of messages that were already completed.

**Pull request merged**